metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,186 @@
+package io.iohk.metronome.checkpointing.service
+
+import cats.data.{NonEmptyList, NonEmptyVector, OptionT}
+import cats.effect.concurrent.{Ref, Semaphore}
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.implicits._
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.models.{
+  Block,
+  CheckpointCertificate,
+  Ledger
+}
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync, N] private (
+    ledgerTree: Ref[F, Map[Block.Hash, (Ledger, Block.Header)]],
+    ledgerUpdateSemaphore: Semaphore[F],
+    maxConcurrentUpdates: Int,
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Boolean] = {
+    {
+      //TODO: Validate transactions PM-3131/3132
+      false.pure[F]
+    } >>
+      ledgerUpdateSemaphore.withPermit {
+        updateLedger(block).map {
+          _.hash == block.header.postStateHash
+        }
+      }
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Unit] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+    if (commitQC.blockHash != block.hash) {
+      //Several options here, depending on how we want to schedule the execution (BlockExecutor):
+      // - disallow this condition
+      // - ignore
+      // - execute block without constructing CC
+      ().pure[F]
+    } else {
+      for {
+        _          <- ledgerUpdateSemaphore.acquireN(maxConcurrentUpdates)
```

Comment:
I guess there are two reasons for locking: to avoid executing the same block twice for validation reasons (but this is a cheap in-memory operation) and, more importantly, to avoid re-entering blocks into the cache when they are no longer on the longest committed path. 

I thought the second case could also be protected against if, at the time of saving the result into the tree, we checked that it still has the parent in the tree. If it's been cleared out, then we don't need its descendants either.

Reply:
> at the time of saving the result into the tree, we checked that it still has the parent in the tree. If it's been cleared out, then we don't need its descendants either.

Possibly that could get rid off locking. I think I'd just need to change `updateLedger` so that it reads the `ledgerTree` map only once for the whole traversal (instead of for each step). WDYT?

Reply:
I admit I find it difficult to think though the locking. Currently you allow 4 validations/creations that can all add to the state, but require exclusive access for execution. So there is a theoretical potential for double entry because if we have a chain like this:
```
        D
      / 
A - B 
      \ 
       C
```
And we try to validate `C` and `D` at the same time, and both realise `B` is not in the ledger, they will both concurrently fetch the state of `A`, apply `B` and enter it into the state. So that's not a reason to limit concurrency then.

That leaves the question of removing data from the state and not re-entering it. 

This is a bit tricky because we could commit an interim value when you already have calculated the state for descendants. I would use locking there to check that any state I re-enter as a result of the traversal is not older than the potentially updated last executed state. Not that it would be a big deal if you re-entered either since you clear out everything every time.

Reply:
I don't this is relevant anymore with locking gone. If so, please resolve this convo.